### PR TITLE
MEPTS-336: Updating queries according to the new specifications

### PR DESCRIPTION
**Pull request by rubailly**
The system will select the source with the most recent registration date (between ART Pick up date, date of registration of
Tipo de Levantamento and date of registration of Modelo diferenciado (Dispensa Trimestral or Dispensa Semestral) prior to
the reporting end date. For patients who have more than one source on the same most recent ART pick up date, consult
date with tipo de Levantamento or Dispensa Trimestral/Dispensa Semestral, the system will use only FILA to determine
patients on each ARV Dispensation Interval. if the patients has tipo de Levantamento or Modelos Diferenciados (Dispensa
Trimestral/Dispensa Semestral) on the same consultation date, the sytem will use the Tipo de Levantamento to determine
the patients on each ARV dispensation Interval.
Patients will only be counted once in each disaggregation.

- For <3 months of ARVs dispense to active patient’s on ART as:

o All active patients whose next ART pick-up is scheduled for <83 days after the date of their last ART drug pick-up
(Fila) or marked as DM on Ficha Clinica Mastercard on last Tipo de Levantamento

- For 3-5 months of ARVs dispensed to patient’s on ART as:

o All active patients whose next ART pick-up is scheduled for 83-173 days after the date of their last ART drug pick-up
(FILA) or marked as DT on Last Tipo de Levantamento on Ficha Clinica Mastercard
▪
or Last “Dispensa Trimestral (DT)” as Iniciar (I) or Manter (C)

- For 6 or more months of ARVs dispensed to active patient’s on ART:

o All active patients whose next ART pick-up is scheduled for >173 days after the date of their last ART drug pick-up
(FILA) or marked as DS on Last Tipo de Levantamento or Ficha Clinica Mastercard
or Last “Dispensa Semestral (DS)” as Iniciar (I) or Manter (C)

If the patient has last ​ QUARTERLY DISPENSATION (DT) or SEMESTRAL DISPENSATION (DS) with value COMPLETED
(id=1267) it should be retired from this indicator regardless of any other indication of being active on ART

### CODE REVIEW

**Comment by ynurmahomed** [comment]
@rubailly please format the code before submitting the PR.

`mvn com.coveo:fmt-maven-plugin:format` or build the project 

SQL strings should also be formatted

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java [hunk 1]
```diff
@@ -841,6 +790,30 @@ private CohortDefinition getPatientsWithMonthlyTypeOfDispensation() {
     cd.addValue(hivMetadata.getMonthlyConcept());
     return cd;
   }
+  
+  public SqlCohortDefinition getPatientsWithMonthlyTypeOfDispensation(List < EncounterType > encounterTypes) {
```
Comment (ynurmahomed) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java]: If encounter types are always going to be:
```
    Arrays.asList(
            hivMetadata.getARVPharmaciaEncounterType(),
            hivMetadata.getAdultoSeguimentoEncounterType());
```
there is no reason to use parameters and a for loop, please inline them in the query and use String.format like all the other queries in the project.

**Review by ynurmahomed — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java [hunk 2]
```diff
@@ -842,6 +794,41 @@ private CohortDefinition getPatientsWithMonthlyTypeOfDispensation() {
...
 
+  public SqlCohortDefinition getPatientsWithMonthlyTypeOfDispensation(
+      List<EncounterType> encounterTypes) {
+    SqlCohortDefinition query = new SqlCohortDefinition();
+    StringBuilder queryStr = new StringBuilder();
+    queryStr.append(
+        "select pp.patient_id from (select lst.patient_id, lst.encounter_datetime from (select last_encounter.patient_id, last_encounter.encounter_id, last_encounter.encounter_datetime from (select e.patient_id, e.encounter_datetime, e.encounter_id from encounter e where e.encounter_type in(");
+
+    int i = 0;
+    for (EncounterType et : encounterTypes) {
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java]: I think we need to be consistent in our approaches, all other queries we kept the encounterIds in line not in a loop, I suggest to use the same approach here
@rubailly

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java [hunk 3]
```diff
@@ -842,6 +794,41 @@ private CohortDefinition getPatientsWithMonthlyTypeOfDispensation() {
...
+      queryStr.append(et.getEncounterTypeId());
+
+      i++;
+    }
+
+    queryStr.append(
+        ") and e.encounter_datetime <= :onOrBefore and e.location_id = :location and e.voided=0 order by e.encounter_datetime desc, case when e.encounter_type = "
+            + encounterTypes.get(0).getEncounterTypeId()
+            + " then 1 else 2 end ) as last_encounter group by last_encounter.patient_id) as lst, obs o where lst.encounter_id=o.encounter_id and o.voided=0 and (( o.concept_id="
+            + hivMetadata.getReturnVisitDateForArvDrugConcept().getConceptId()
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java]: @rubailly Also use the % notation for variables to pass on queries

**Review by psmatsinhe — CHANGES_REQUESTED** [review]

**Review by psmatsinhe — APPROVED** [review]

**Pull request merged**